.circleci/build-workbase-win/build-workbase-win.py

```diff
@@ -0,0 +1,166 @@
+subprocess.check_call([
+    'gcloud', 'config', 'set', 'compute/zone', 'europe-west1-b'
+])
+
+print('Generating password for instance')
+instance_password = subprocess.check_output([
+    'openssl', 'rand', '-base64', '12'
+]).strip()
+
+instance_name = 'circleci-{}-{}'.format(
+    os.getenv('CIRCLE_JOB'), os.getenv('CIRCLE_BUILD_NUM'))
+
+print('Generating bootup script for instance [{}]'.format(instance_name))
+with tempfile.NamedTemporaryFile(suffix='.ps1') as powershell_script:
+    with open('.circleci/build-workbase-win/instance-setup-template.ps1') as template:
+        powershell_script.write(template.read().replace('INSTANCE_PASSWORD', instance_password).encode())
+        powershell_script.flush()
+
+    print('Provisioning instance [{}]'.format(instance_name))
+    subprocess.check_call([
+        'gcloud', 'compute', 'instances', 'create', instance_name,
+        '--image-project', 'windows-cloud', '--image-family', 'windows-2019',
+        '--machine-type', 'n1-standard-1', '--metadata-from-file',
+        'sysprep-specialize-script-ps1={}'.format(powershell_script.name)
+    ])
+
+print('Storing instance\'s external IP')
+instance_ip = subprocess.check_output([
+    'gcloud', '--format', 'value(networkInterfaces[0].accessConfigs[0].natIP)',
+    'compute', 'instances', 'list', '--filter', 'name={}'.format(instance_name)
+])
+
+
+try:
+    print('Waiting for instance to some alive (sshd)')
+    wait_for_ssh(instance_ip, 'circleci', instance_password)
+
+    print('Executing command remotely')
+    ssh('dir', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('Executing PowerShell command remotely [Get-LocalUser]')
+    ssh('powershell Get-LocalUser', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('Installing git')
+    ssh('choco install git -y', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('Installing nodejs 8')
+    ssh('choco install nodejs -y --version 8.15.0', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('Cloning workbase')
+    ssh(
+        'refreshenv && git clone {repo_url} repo && cd repo && git checkout {repo_commit}'.format(
+            repo_url=replace_git_url_to_https(os.getenv('CIRCLE_REPOSITORY_URL')),
+            repo_commit=os.getenv('CIRCLE_SHA1')),
+        instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('[Remote]: npm install')
+    ssh('refreshenv && cd repo && npm install', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('[Remote]: npm run build')
+    ssh('refreshenv && cd repo && npm run build', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('Copying built Workbase executable from remote to local')
+    scp('C:\\Users\\circleci\\repo\\build\\GRAKNW~1.EXE', './GRAKN.exe', instance_ip, 'circleci', instance_password, check_call=True)
+
+    print('Verifying local file')
+    subprocess.check_call(['file', './GRAKN.exe'])
```

Comment:
Why is `GRAKN.exe` capital?

Reply:
I think `ssh` and `scp` should use `subprocess.check_call()` internally. They should always throw an exception. The main code would then be wrapped in a try / catch / finally:

```
def scp(...):
  subprocess.check_call(...)

def ssh(...):
  subprocess.check_call(...)

try:
  ...
except subprocess.CalledProcessException as e:
  ...
finally:
 ...

Reply:
and the reason for that is consistency, so the user don't have to think about whether a specific system call would throw an exception or return non-zero on failure (which takes a bit of cognitive effort).

You supply `check_call=True` in most of the invocations anyway, except in `wait_for_ssh`.

Reply:
And in `wait_for_ssh` you would just have to catch them:
```
def wait_for_ssh(...):
  try:
    ssh(...)
  catch subprocess.CalledProcessException as e:
    // timeout
  ...
```

Reply:
@lolski 

>Why is `GRAKN.exe` capital?

No specific reason, fixed in https://github.com/graknlabs/workbase/pull/7/commits/d379eab3f3ee10161e1c5dd6da24ccb9d2f4e268

Reply:
@lolski 

>And in `wait_for_ssh` you would just have to catch them:

Using try-except in place of control flow statements is considered a bad practice. Instead, in https://github.com/graknlabs/workbase/pull/7/commits/44521c357c51986a0c08c6d6cdcd70a9896ca8a7 I introduced `check_call` wrapper which turns a function calling `subprocess.check_call` (i.e. `scp`,`ssh`) to a function that calling `subprocess.call` — that is, return command's exit code instead of throwing